File: src/hypomnema/ontology/normalizer.py
"""Canonical string normalization and synonym resolution."""

from __future__ import annotations

import json
import re
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from hypomnema.llm.base import LLMClient


def normalize(name: str) -> str:
    """Normalize entity name: strip, collapse whitespace, lowercase, strip trailing punctuation.

    Raises:
        ValueError: If result is empty after normalization.
    """
    result = re.sub(r"\s+", " ", name.strip()).lower().rstrip(".,;:!?")
    if not result:
        raise ValueError("Entity name is empty after normalization")
    return result


_SYNONYM_SYSTEM = (
    "You are a terminology normalizer. Given a JSON list of entity names, "
    "group synonyms and pick one canonical name per group. "
    'Return ONLY valid JSON: {"mapping": {"original": "canonical", ...}}. '
    "Every input name must appear as a key. Names that aren't synonyms map to themselves."
)


async def resolve_synonyms(
    llm: LLMClient,
    names: list[str],
) -> dict[str, str]:
    """Ask LLM to merge synonymous names within a batch.

    Returns:
        Dict mapping each input name to its canonical form.
    """
    if len(names) <= 1:
        return {n: n for n in names}

    prompt = f"Normalize these entity names:\n{json.dumps(names)}"
    result = await llm.complete_json(prompt, system=_SYNONYM_SYSTEM)
    mapping = result.get("mapping", {})

    # Ensure every input name has a mapping (fall back to identity)
    return {n: mapping.get(n, n) for n in names}

File: server/internal/ontology/normalizer.go
package ontology

import (
	"context"
	"fmt"
	"strings"
	"unicode"

	"github.com/junhewk/hypomnema/internal/llm"
)

// NormalizeName applies deterministic normalization: lowercase, collapse whitespace, strip trailing punctuation.
func NormalizeName(name string) string {
	name = strings.TrimSpace(name)
	name = collapseWhitespace(name)
	name = strings.ToLower(name)
	name = strings.TrimRightFunc(name, func(r rune) bool {
		return unicode.IsPunct(r) && r != ')' && r != ']'
	})
	return name
}

// ResolveSynonyms asks the LLM to group synonyms and pick canonical forms.
func ResolveSynonyms(ctx context.Context, client llm.Client, names []string) (map[string]string, error) {
	if len(names) <= 1 {
		m := make(map[string]string)
		for _, n := range names {
			m[n] = n
		}
		return m, nil
	}

	prompt := fmt.Sprintf(`Group these entity names by synonymy and pick one canonical form per group.
Return JSON: {"mapping": {"original_name": "canonical_name", ...}}

Names: %v`, names)

	system := "You normalize entity names. Group synonyms, pick the most precise canonical form. Preserve the original language of each name — do NOT translate. For mixed-language names, keep each span in its original language. Return only the JSON mapping."

	result, err := client.CompleteJSON(ctx, prompt, system)
	if err != nil {
		return nil, err
	}

	mapping := make(map[string]string)
	if raw, ok := result["mapping"].(map[string]any); ok {
		for k, v := range raw {
			if s, ok := v.(string); ok {
				mapping[k] = NormalizeName(s)
			}
		}
	}

	// Ensure all input names have a mapping
	for _, n := range names {
		if _, ok := mapping[n]; !ok {
			mapping[n] = n
		}
	}

	return mapping, nil
}

func collapseWhitespace(s string) string {
	var b strings.Builder
	space := false
	for _, r := range s {
		if unicode.IsSpace(r) {
			if !space {
				b.WriteRune(' ')
				space = true
			}
		} else {
			b.WriteRune(r)
			space = false
		}
	}
	return b.String()
}
